provider/gpkg/gpkg_test.go

```diff
@@ -0,0 +1,418 @@
+			return
+		}
+
+		var featureCount int
+		err = p.TileFeatures(context.TODO(), tc.layerName, &tc.tile, func(f *provider.Feature) error {
+			featureCount++
+			return nil
+		})
+		if err != nil {
+			t.Errorf("err fetching features: %v", err)
+			return
+		}
+
+		if tc.expectedFeatureCount != featureCount {
+			t.Errorf("expected %v got %v", tc.expectedFeatureCount, featureCount)
+			return
+		}
+	}
+
+	tests := map[string]tcase{
+		// roads_lines bounding box is: [23.6655, 37.85, 23.7958, 37.9431] (see gpkg_contents table)
+		"tile outside layer extent": tcase{
+			config: map[string]interface{}{
+				"filepath": GPKGAthensFilePath,
+				"layers": []map[string]interface{}{
+					{"name": "rd_lines", "tablename": "roads_lines"},
+				},
+			},
+			layerName: "rd_lines",
+			tile: MockTile{
+				srid: tegola.WGS84,
+				bufferedExtent: [2][2]float64{
+					{20.0, 37.85},
+					{23.6, 37.9431},
+				},
+			},
+			expectedFeatureCount: 0,
+		},
+		// rail lines bounding box is: [23.6828, 37.8501, 23.7549, 37.9431]
+		"tile inside layer extent": tcase{
+			config: map[string]interface{}{
+				"filepath": GPKGAthensFilePath,
+				"layers": []map[string]interface{}{
+					{"name": "rl_lines", "tablename": "rail_lines"},
+				},
+			},
+			layerName: "rl_lines",
+			tile: MockTile{
+				srid: tegola.WGS84,
+				bufferedExtent: [2][2]float64{
+					{23.6, 38.0},
+					{23.8, 37.8},
+				},
+			},
+			expectedFeatureCount: 187,
+		},
+		"zoom token": tcase{
+			config: map[string]interface{}{
+				"filepath": GPKGNaturalEarthFilePath,
+				"layers": []map[string]interface{}{
+					{
+						"name": "land1",
+						"sql": `
+							SELECT
+								fid, geom, featurecla, min_zoom, 22 as max_zoom, minx, miny, maxx, maxy
+							FROM
+								ne_110m_land t JOIN rtree_ne_110m_land_geom si ON t.fid = si.id
+							WHERE
+								!BBOX! AND min_zoom <= !ZOOM!`,
+					},
+				},
+			},
+			layerName: "land1",
+			tile: MockTile{
+				Z:    1,
+				srid: tegola.WebMercator,
+				bufferedExtent: [2][2]float64{
+					{-20026376.39, -20048966.10},
+					{20026376.39, 20048966.10},
+				},
+			},
+			expectedFeatureCount: 101,
+		},
+		"zoom token 2": tcase{
+			config: map[string]interface{}{
+				"filepath": GPKGNaturalEarthFilePath,
+				"layers": []map[string]interface{}{
+					{
+						"name": "land2",
+						"sql": `
+							SELECT
+								fid, geom, featurecla, min_zoom, 22 as max_zoom, minx, miny, maxx, maxy
+							FROM
+								ne_110m_land t JOIN rtree_ne_110m_land_geom si ON t.fid = si.id
+							WHERE
+								!BBOX! AND min_zoom <= !ZOOM! AND max_zoom >= !ZOOM!`,
+					},
+				},
+			},
+			layerName: "land2",
+			tile: MockTile{
+				Z:    0,
+				srid: tegola.WebMercator,
+				bufferedExtent: [2][2]float64{
+					{-20026376.39, -20048966.10},
+					{20026376.39, 20048966.10},
+				},
+			},
+			expectedFeatureCount: 44,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
```

Comment:
You need to add:
`tc := tc`
this is to make sure that a copy of tc is made, otherwise, it going to all be the same.